### DISCUSSION TITLE
Spread Operator Like JS, to Create Different Objects (Especially Extended Objects) Easily

### DISCUSSION THREAD

**Discussion by sahin52** [dotnet/csharplang#6971] [Language Ideas] [answered]
Hi, in JS language we have a spread operator, so that we can create an object from another, it takes all properties from the other class.
for example:
```js
let elements = [{"n":1},{"n":5}]
let y = elements.map((item,i)=>{...item, "description":"current: "+i})
```
I could easily create another object.
Similarly in C#, I have two classes one of them is base class, and the other one extends it. I can't create an extended class instance, by copying every other thing that exists in base class.
```csharp
public class Base{
    public string Name {get;set;}
    public string ID {get;set;}
// a bunch of lots of other attributes
}
public class Extended: Base{
     public string Description {get;set;}
}
```
I just want to create an extended class instance from a base class instance. I can copy every attribute, I can loop through all attributes but this is so boring. Can we create something for this? Like for example this:
```csharp
var baseList = new List<Base>();
baseList.Add(new Base(){ID="1",Name="name"})
//baseList.Add....
var extendedList = baseList.Select(base=>new Extended(){...base, Description="it's extended"}).ToList()
```

**Comment by En3Tho** [comment]
I believe there are downsides to this: this is property based like get/set or get/init so there is no way to guarantee correctness of object creation.

Constructors are hard to link with this feature. Private fields too.

What if derived class declares a "new" member with same name but it's get-only?

Maybe if you narrow it down to records only it can work? Not sure

Also when you think about ts/js spread then obviously structural spread comes to mind. Like why would you tie yourself to inheritance with this feature? Seems limited.

**Comment by sahin52** [comment]
Here is the problem I have: https://stackoverflow.com/questions/14613919/copying-the-contents-of-a-base-class-from-a-derived-class
I am currently using Map method given in the answers. I am using this for DTO classes, they don't have a very important constructor or private members.
I will use Map if this won't be provided by C#, constuctors and other stuff already are a problem if they are a problem for my topic.

**Reply by En3Tho** [reply]
I think that your best choice right now is
1) use records as they provide a structural equality out of the box and easier to work with as DTO's. Also with expression might be good for your case as it is somewhat similar to spread operator (at least when types are same)
2) Write a souce generator which will automatically write methods you need. (like extension method
 ```csharp
Extended ToExtended(this Base base) { ... }
```

As for the idea: I perosnally think it might be good and useful in some cases. It's just that there are lots of caveats which might limit this feature to a very small subset of what is possible in TS/JS

**Comment by svick** [comment, marked as answer]
This is a duplicate of https://github.com/dotnet/csharplang/discussions/2879.
